state/fork/terminals.go

```diff
@@ -0,0 +1,64 @@
+package fork
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/storage"
+)
+
+func IncludingBlock(lowestBlockId flow.Identifier) Terminal {
+	return terminalBlock{lowestBlockId, true}
+}
+
+func ExcludingBlock(lowestBlockId flow.Identifier) Terminal {
+	return terminalBlock{lowestBlockId, false}
+}
+
+type terminalBlock struct {
+	terminalBlockID flow.Identifier
+	inclusive       bool
+}
+
+func (t terminalBlock) Translate2Height(headers storage.Headers) (uint64, sanityCheckLowestVisitedBlock, error) {
+	terminalHeader, err := headers.ByBlockID(t.terminalBlockID)
+	if err != nil {
+		return 0, noopSanityChecker, fmt.Errorf("failed to retrieve header of terminal block %x: %w", t.terminalBlockID, err)
+	}
```

Comment:
```suggestion
func (t terminalBlock) Translate2Height(headers storage.Headers) (uint64, error) {
	terminalHeader, err := headers.ByBlockID(t.terminalBlockID)
	if err != nil {
		return 0, noopSanityChecker, fmt.Errorf("failed to retrieve header of terminal block %x: %w", t.terminalBlockID, err)
	}
	
	terminalHeaderByHeight, err := headers.ByHeight(terminalHeader.Height)
	// error check
	if terminalHeader.ID() != terminalHeaderByHeight.ID() {
		// throw error
	}
```

I think it's fine to have these kinds of sanity checks, but increasing the complexity of the API to accomodate them feels like a bridge too far for me 😅. In this case I think we can limit the scope to this method as suggested above.

In general though I feel it makes more sense to put state integrity checks like this in the codepath **where the state could be changing** in way that breaks an assumption (ie. when we insert the `height->blockID` index) rather than at all the points where we read the state.

Reply:
On the one hand, I totally agree that the API would be much cleaner, if we removed the `sanityCheckLowestVisitedBlock`. On the other hand, this check covers some distinct failure cases, that would otherwise return the implementation with the wrong result (instead of error'ing). Specifically:
* Consider the case where we have two forks:
   ``` 
   A <- B <- C <- D
     ^- X <- Y <- Z 
  with height:
   10   11   12   13
   ``` 
* I now call `TraverseBackward(headers, `**`D`**`, visitor, IncludingBlock(`**X**`))`. The business logic translates block `X` to height `11` and then traverses the fork with head `D` backwards until it has reached block with height `11`. As `D` and `X` are on different forks and the traversal with end with `B` as the last block; but does _not_ include `X`. This is because the inputs are inconsistent. 
* A similar edge case of inconsistent inputs is `TraverseBackward(headers, `**`D`**`, visitor, ExcludingBlock(`**X**`))`.

Of course, it is a design option to not cover the this edge case in the framework and leave it up to the application logic:
* increases complexity of the business logic (if edge cases are covered and tested) or its safety (if edge cases are uncovered)
* increases runtime complexity: This is a specific check on the terminal block. Adding this check to the visitor would require the visitor to check if the block it is currently visiting is the terminal block (or the parent). Thereby, we essentially duplicate the evaluation of the termination criterion. 
   * removing the sanity check does not significantly impact runtime complexity: as the check is only called _once_ on the terminal block

Some part of me also wants to remove this and simplify the code. Then, I am wondering, does the current design actually impact the complexity of the logic calling it? 
https://github.com/onflow/flow-go/blob/9986c78f732eae64994985a4ea62d311a0f17197/module/builder/consensus/builder.go#L199 would still look exactly the same when removing the sanity check. 

So here is the logical summary from my perspective: 
* the sanity check makes the framework safer because it allows the framework to error in the case of inconsistent input (as opposed to returning possibly incorrect results)
* the sanity check makes increases internal complexity of the framework 
* the sanity check makes does not impact external API complexity

Reply:
```golang
terminalHeaderByHeight, err := headers.ByHeight(terminalHeader.Height)
```

Sorry, forgot to mention: This only works for finalized blocks. What if you want to traverse unfinalized blocks (or orphaned blocks for whatever reason).